src/main/java/com/google/sps/servlets/SpecificGuideServlet.java

```diff
@@ -0,0 +1,50 @@
+package com.google.sps.servlets;
+
+import java.io.IOException;
+
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import com.google.gson.Gson;
+import com.google.sps.data.Error;
+import com.google.sps.data.Guide;
+import com.google.sps.storage.DatastoreGuides;
+
+@WebServlet("/guides/*")
+public class SpecificGuideServlet extends HttpServlet{
+    static final long serialVersionUID = 0;
+    private final DatastoreGuides helper = new DatastoreGuides();
+    private final Gson gson = new Gson();
+
+    private long getIdFromPath(String path){
+        String[] parts = path.split("/");
+        return Long.parseLong(parts[1]);
+    }
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException{
+        response.setContentType("application/json");
+
+        long id;
+        try{
+            id = getIdFromPath(request.getPathInfo());
+        }catch(NumberFormatException exception){
```

Comment:
Since this error could be thrown in many different places in your code, it's easier to define an error handler that catches and handles any exception of this type. It would also be infeasible to handle every possible error in each of your servlets.

I think this tutorial looks helpful: https://www.tutorialspoint.com/servlets/servlets-exception-handling.htm
https://cloud.google.com/appengine/docs/standard/java/config/webxml#error-handlers

Rather than mapping NumberFormatException, I would map any instance of java.lang.IllegalArgumentException to a 400 error.

You sometimes want to be careful about returning an error string in the response, since it might expose internal or private details unintentionally. At Google, our servers automatically return a stack trace if it was an internal request, and a canned response if it was external. If the response doesn't include enough detail, you would still be able to look at server logs to troubleshoot.

Reply:
Ok, I understand what I have to do, but I don't have clear where the web.xml file is in our project or the equivalent in maven.

Reply:
Good question. I'd start by throwing (or rather, not catching) the error, and seeing what you get in the HTTP response. It's possible you don't really need to implement an error handler.

Reply:
If I don't catch the error, it returns 500 server error, with a stack trace. I guess what I need to do is make the check somewhere else and reuse that.